Avoid leap second update on astropy import

taldcroft opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
http://docs.astropy.org/en/latest/development/workflow/development_workflow.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/workflow/git_edit_workflow_examples.html#astropy-fix-example . -->

<!-- Astropy coding style guidelines can be found here:
https://docs.astropy.org/en/latest/development/codeguide.html#coding-style-conventions
Our testing infrastructure enforces to follow a subset of the PEP8 to be
followed. You can check locally whether your changes have followed these by
running the following command:

tox -e codestyle

-->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

This pull request is to address #11426. Before really getting much further this initial draft raises some questions that probably need @mhvk to answer.

The current version is the minimal implementation that can avoid the leap second update when importing parts of astropy which create Time objects (e.g. coordinates). It moves the leap second checking from `Time.__new__` to `Time._set_scale` (when one of the scales is `'utc'`).

This immediately fails a few of the tests in `test_update_leap_seconds.py`. After some head scratching I discovered that the tests are only passing in `main` because of a what seems like a side-effect of the implementation. The side effect is that `Time` objects are created to represent the `expires` attributes. For instance, one might have decided to represent those `expires` objects as `datetime` instead. I did this with these diffs (in `main`) and indeed saw test failures.
```
diff --git a/astropy/utils/iers/iers.py b/astropy/utils/iers/iers.py
index 01a471750..4f55039fe 100644
--- a/astropy/utils/iers/iers.py
+++ b/astropy/utils/iers/iers.py
@@ -1045,8 +1045,7 @@ class LeapSeconds(QTable):
                 if match:
                     day, month, year = match.groups()[0].split()
                     month_nb = MONTH_ABBR.index(month[:3]) + 1
-                    expires = Time(f'{year}-{month_nb:02d}-{day}',
-                                   scale='tai', out_subfmt='date')
+                    expires = datetime(int(year), month_nb, int(day))
                     break
             else:
                 raise ValueError(f'did not find expiration date in {file}')
@@ -1118,9 +1117,7 @@ class LeapSeconds(QTable):
             with erfa.
         """
         current = cls(erfa.leap_seconds.get())
-        current._expires = Time('{0.year:04d}-{0.month:02d}-{0.day:02d}'
-                                .format(erfa.leap_seconds.expires),
-                                scale='tai')
+        current._expires = erfa.leap_seconds.expires
         if not built_in:
             return current
```
A minimal example, inspired by one of the simpler tests, is below. Again, this only works because the auto-update machinery happens to be called in `from_erfa()`.
```
from astropy.utils import iers
from datetime import datetime

built_in = iers.LeapSeconds.from_iers_leap_seconds()
erfa_ls = iers.LeapSeconds.from_erfa()
print('assert', erfa_ls.expires > datetime.now())
```

In the version here in this PR, the side effect goes away and so the tests fail.

Questions:
- Suggestions for where to go with this issue?
- Will the complexity of the code being potentially re-entrant go away now? I confess to begin unable to follow the code logic because of the leap second handling code calling `Time` itself.

Also, in #11426 we settled on a solution with a new configuration variable `leap_second_check_action`. In retrospect this seems a bit over-engineered and I wonder if we can avoid the extra complexity by just relying on `auto_download` to disable leap second internet checking. This goes along with the big-picture goal of trying to reduce complexity (knobs) in astropy.
 
<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #<Issue Number>

eteq commented:
I'm going to flag this as `whast-new-needed` because if we can get this working for 4.3 it's definitely worth putting it and the related fixes in the what's new

taldcroft commented:
Not a blocker, but several people have chimed in through various channels that this would be quite desirable.  @mhvk - I know it's late but do you have time to weigh in?

taldcroft commented:
I suspect I can get tests passing if the approach is considered OK.

mhvk commented:
@taldcroft - yes, I think the general approach is right. When you wrote about problems with the expiry date being a `Time`, I recall that I had the same problem in earlier iterations, and tried to avoid importing it, using `datetime`, etc. But `Time` is used elsewhere in `utils.iers` as well as in `Table` (which `IERS` and the leap second table are based on), so it became a nightmare to avoid it. With the current scheme, it may be possible to still use `Time` for the expiration date, e.g., by using `format='tai'` for it and be careful with getting "now", or by only comparing `jd1` instead of the full time object.

Aside: I will try to look over the PR, though this week is quite stressed already; I would very much like to have it in 4.3 given the problems that my implementation has caused. And I think it can fairly be considered a bug-fix...

taldcroft commented:
Thanks, I'll see if I can get tests passing and get something reasonable and easy to review.

taldcroft commented:
@mhvk - assuming this gets through CI, this is ready for final review. The code is just the same as before but I updated some docs. 

The only interesting or controversial bit is about testing, where I added an `assert` check that gets run during test collection. The idea is to test (and maintain) that astropy can be imported without triggering a download of a non-local leap second file. I *think* that the new code actually checks that but you need to have a look.

Otherwise I think this should be a quick review. 🤞

taldcroft commented:
Dang, my new test already failed. Sigh, will look.

mhvk commented:
The changes themselves look good, and I guess at least now we know that your check does work...  I'll also try to have a look, but I first have to get an annual report in...

taldcroft commented:
I found the culprit, update coming.

mhvk commented on astropy/timeseries/tests/test_downsample.py:
```diff
@@ -12,8 +12,6 @@
 
 INPUT_TIME = Time(['2016-03-22T12:30:31', '2016-03-22T12:30:32',
                    '2016-03-22T12:30:33', '2016-03-22T12:30:34'])
-ts = TimeSeries(time=INPUT_TIME, data=[[1, 2, 3, 4]], names=['a'])
```
Good catch. Now see another benefit of doing setup inside `setup_class`, etc.

taldcroft replied:
I had to basically monkeypatch erfa to force an exception and find this. In this case `setup_class` is not strictly needed since it was just one function, but I have been paying attention to your style of testing...

mhvk commented on astropy/utils/iers/tests/test_leap_second.py:
```diff
@@ -13,6 +13,63 @@
 from astropy.utils.iers import iers
 from astropy.utils.data import get_pkg_data_filename
 
+# Import every top-level astropy subpackage as a test that the ERFA leap second
```
How about putting this in a new test file? E.g., `astropy/tests/tests` has a `test_imports.py` and one could add an `test_no_net_on_import.py` (or better name...) there. Then, if someone does something like what I did for the leap seconds, they'll see a failure in a file that will hopefully make sense.

taldcroft replied:
I actually did this initially. The problem I ran into is that the current `test_leap_second.py` requires the modified `erfa.leap_seconds` to pass, whereas the new imports-only test file would require the unmodified version to pass. So the success of tests passing depends on the order of test collection. 

The only ways I could see around this are:
1. Put things in one file where the sequence of testing `erfa.leap_seconds` and then updating it is controlled, or
1. Make some of the tests here have some kind of setup that does the leap second update. IIRC there are 3 tests here that fail without it, so I guess that is an option, but the first option seemed a bit more straightforward.

mhvk replied:
That makes sense. One option to at least make this shorter is to do
```
from astropy.tests.tests import test_imports

test_imports()
# Instead of all the explicit imports
```

mhvk commented on astropy/utils/iers/tests/test_leap_second.py:
```diff
@@ -13,6 +13,63 @@
+import astropy.visualization
+import astropy.wcs
+
+# Now test that the erfa leap_seconds table has not been udpated. This must be
```
We could still leave this test here, but perhaps cleaner to move it to the new file, and explain there that it is a regression test for a particular example.

mhvk reviewed (commented):
This is great! I'm impressed that it ended up so (relatively!) straightforward. Possibly, having split out `erfa` has required some cleanup that made this easier. In any case, thanks for insisting this was important *and* fixing it!!!

Only one comment on the location of the tests.

mhvk reviewed (approved):
@taldcroft - an inline suggestion to simplify the tests.

I see your point that even for test collection it may be tricky to know the order. I wouldn't mind playing around with it, but may not get to it today. So, let me approve this now, so that if I don't, it can be merged by the end of the day.

Overall, again very happy that the solution ended up being so simple & clean!

taldcroft commented:
Ah, I didn't know about the test imports functionality. I'll do that.
